carstore/bs.go

```diff
@@ -947,39 +947,54 @@ func (cs *CarStore) Stat(ctx context.Context, usr models.Uid) ([]UserStat, error
 }
 
 func (cs *CarStore) TakeDownRepo(ctx context.Context, user models.Uid) error {
-	var shards []CarShard
+	var shards []*CarShard
 	if err := cs.meta.Find(&shards, "usr = ?", user).Error; err != nil {
 		return err
 	}
 
-	for _, sh := range shards {
-		if err := cs.deleteShard(ctx, &sh); err != nil {
-			if !os.IsNotExist(err) {
-				return err
-			}
+	if err := cs.deleteShards(ctx, shards); err != nil {
+		if !os.IsNotExist(err) {
+			return err
 		}
 	}
 
-	if err := cs.meta.Delete(&CarShard{}, "usr = ?", user).Error; err != nil {
-		return err
-	}
-
 	return nil
 }
 
-func (cs *CarStore) deleteShard(ctx context.Context, sh *CarShard) error {
+func (cs *CarStore) deleteShards(ctx context.Context, shs []*CarShard) error {
 	ctx, span := otel.Tracer("carstore").Start(ctx, "deleteShard")
 	defer span.End()
 
-	if err := cs.meta.Delete(&CarShard{}, "id = ?", sh.ID).Error; err != nil {
+	var ids []uint
+	for _, sh := range shs {
+		ids = append(ids, sh.ID)
+	}
+
+	txn := cs.meta.Begin()
+
+	if err := txn.Delete(&CarShard{}, "id in (?)", ids).Error; err != nil {
 		return err
 	}
 
-	if err := cs.meta.Delete(&blockRef{}, "shard = ?", sh.ID).Error; err != nil {
+	if err := txn.Delete(&blockRef{}, "shard in (?)", ids).Error; err != nil {
 		return err
 	}
 
-	return cs.deleteShardFile(ctx, sh)
+	if err := txn.Commit().Error; err != nil {
+		return err
+	}
+
+	var outErr error
+	for _, sh := range shs {
+		if err := cs.deleteShardFile(ctx, sh); err != nil {
+			if !os.IsNotExist(err) {
+				return err
+			}
+			outErr = err
+		}
+	}
+
+	return outErr
```

Comment:
Is this bit just to return at least one `IsNotExist` if any of the shards end up already having been deleted but we don't wanna let that stop us from deleting the others, however if there's some other error we're interested in stopping immediately and returning?

Reply:
exactly, yeah. i want to at least try to delete everything, if we fail because the file is already gone, we continue (but still let our caller know it happened). if we fail for some other error i think it makes sense to bail out immediately